RxSwift/Subjects/ReplaySubject.swift

```diff
@@ -59,7 +59,7 @@ public class ReplaySubject<Element>
     }
     
     /// Returns observer interface for subject.
-    public func asObserver() -> SubjectObserverType {
```

Comment:
Required as otherwise when Xcode12 is looking at the generated .swiftinterface it sees that SubjectObserverType = ReplaySubject<Element> and gets confused with it's a circular reference.

Looking at all the other subjects, they look to have moved away from SubjectObserverType (which is marked deprecated); This was the only one left so presumably it is safe

Reply:
This is a bit tricky since it's a source breaking change. I don't think we'll be able to merge it in because you switched away from an abstracted observer to the concrete type and some people might count on the abstraction. 

Perhaps we could trying AnyObserver here? What do we do in other subjects?

Reply:
ReplaySubject declares that SubjectObserverType is aliased to `ReplaySubject<Element>` already, so the return type of `asObserver()` is always `ReplaySubject<Element>` before or after the change.

I looked at `BehaviourSubject`
https://github.com/ReactiveX/RxSwift/blob/a580d07ed002217fd91d8446c3a852486e9beefa/RxSwift/Subjects/BehaviorSubject.swift#L148

`AsyncSubject` https://github.com/ReactiveX/RxSwift/blob/a580d07ed002217fd91d8446c3a852486e9beefa/RxSwift/Subjects/AsyncSubject.swift#L145

and `PublishSubject`
https://github.com/ReactiveX/RxSwift/blob/a580d07ed002217fd91d8446c3a852486e9beefa/RxSwift/Subjects/PublishSubject.swift#L128

and those all return their own concrete types from `asObserver`. They are final though and ReplaySubject is not, so yeah this would be source-breaking if anyone subclassed ReplaySubject and then called asObserver()
Are we trying to get covariant return types here?

A github search doesn't show any classes that subclass ReplaySubject other than a unit test? Is subclassing ReplaySubject a design goal?

Re: changing - If we changed it to `AnyObserver` then that seems  worse than returning `ReplaySubject`?
I.e. if we return `ReplaySubject` from a subclass, we lose some type information, but if we return `AnyObserver` then we lose *more* type information, and don't seem to benefit here?

OTOH this could also just be a bug in swift, it's hard to know what should and shouldn't work

Reply:
Ah I see. I wasn’t near the code when commenting, I think this fix is fine in this case.
Let’s see what Apple has to say about the internal initializers

Reply:
Thanks. If there is a resolution for the internal initalizers, shall I make ReplaySubject sealed here to match the other subjects? Or leave it as overridable and just change the asObserver method?

Reply:
I'd say we can make it final as well given its a relatively niche subject, and in favor of proceeding into binary stability :)

Reply:
Thanks, I'll make this change and update the PR, just waiting on that swiftinterface thing

Reply:
re-reviewed against the latest commits